### PR TITLE
Migrate build to GitHub actions

### PR DESCRIPTION
**Pull request by seglo**
TODO
- [x] Check, Build, Test
- [x] Whitesource (using `release` environment for secrets, can put `publish` secrets there too, see whitesource workflow for reference)
- [ ] Publish
- [x] Dependency cache
- [x] Print log on failure

### CODE REVIEW

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          git checkout scratch
+
+      - name: Set up JDK 11
+        uses: olafurpg/setup-scala@v10
+        with:
+          java-version: adopt@1.11.0-9
+
+      - name: "Code style check and binary-compatibility check"
+        # Run locally with: sbt 'verifyCodeStyle ; mimaReportBinaryIssues'
+        run: sbt "; verifyCodeStyle; mimaReportBinaryIssues"
```
Comment (ennru) [inline .github/workflows/check.yml:35]: Using the thin sbt client will shorten multiple sbt starts:
```suggestion
        run: sbt --client "verifyCodeStyle; mimaReportBinaryIssues"
```

Comment (seglo) [reply .github/workflows/check.yml]: :+1:

File: .github/workflows/check-build-test.yml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          # FIXME: since we don't know what commit the last tag was it's safer to get entire repo so previousStableVersion resolves
+          fetch-depth: 0
+
+      # FIXME: removed in favour of fetch-depth defined in Checkout step ^. i tried an arbitrarily large fetch depth 1000 and it still didn't work.
+      #      - name: Fetch tags
+      #        run: git fetch --depth=1000 origin +refs/tags/*:refs/tags/*
```
Comment (seglo) [inline .github/workflows/check-build-test.yml:21]: @ennru Any thoughts?

File: .github/workflows/check-build-test.yml [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+      - name: Checkout GitHub merge
+        if: github.event.pull_request
+        run: |-
+          git fetch origin pull/${{ github.event.pull_request.number }}/merge:scratch
+          git checkout scratch
```
Comment (seglo) [inline .github/workflows/check-build-test.yml:27]: I copied this blindly. I'm not sure what the problem with the `checkout` step default behaviour of using a merge commit.

There's also this:

https://github.com/actions/checkout#checkout-pull-request-head-commit-instead-of-merge-commit

Comment (ennru) [reply .github/workflows/check-build-test.yml:27]: In other repos we saw that the `checkout@v2` action didn't do what the documentation said.
The hack above was identified by @chbatey.

Comment (chbatey) [reply .github/workflows/check-build-test.yml:24]: It doesn't use merge commit, it is a lie, @patriknw discovered this and i re-discovered it

Comment (seglo) [reply .github/workflows/check-build-test.yml:24]: Thanks.

**Comment by seglo** [comment]
Recreating on upstream repo to test publishing.

**Pull request closed without merging**